spec/compiler/type_inference/global_spec.cr

```diff
@@ -185,6 +185,20 @@ describe "Global inference" do
       )) { types["Foo"] }
   end
 
+  it "doesn't infer from new if generic" do
+    assert_error %(
+      class Foo(T)
+        def self.new
+          10
+        end
+      end
+
+      $x = Foo.new
+      $x
+      ),
+      "undefined global variable '$x'"
```

Comment:
@jhass The test is actually testing that if the compiler found `Foo.new`, and `Foo` is a generic type, nothing can be inferred from that assignment because right now variables can't have a non-instantiated generic type (that is, `$x`'s type can't be just `Foo`, it must be `Foo(Int32)`, `Foo(Float64)`, etc.). However, if I don't provide a `new` method the compiler will first complain that it can't deduce the generic type for `T` by using the default constructor. So I cheated and provided a `new` method (it can return any type actually), so I could test that the type wasn't inferred from that assignemnt. It's a pretty contrived and non-realistic example.

In a more realistic scenario, if `Foo` is non generic and you override `new` to return, say, an `Int32`, the compiler will say "global variable '$x' must be Foo, not Int32" with the error pointing at the assignment, which would be much more understandable.

There's also the possibility of restricting `T.new` to always return something of type `T`, but we are not sure yet.

In any case, I believe the error messages can be much better than right now, by for example showing the list of assignments and saying what the compiler inferred from each, if it could (and why if it couldn't), when such error happens.